doc/man/gnunet-directory.1
.TH gnunet-directory "1" "30 Mar 2004" "GNUnet"
.SH NAME
gnunet-directory \- display directories and show or delete lists of encountered file identifiers

.SH SYNOPSIS
.B gnunet-directory
[\fIOPTIONS\fR] FILENAMES
.SH DESCRIPTION
.PP
gnunet-directory lists the contents of a GNUnet directory.  It also can be used to manipulate the file-identifier database which is used by GNUnet for building directories.
.PP
gnunet-directory will always list the contents of the GNUnet directories that are passed as filenames.
.PP
Manipulating the file-identifier database is done by passing additional options to gnunet-directory.  Note that by default GNUnet does not build the file-identifier database and the database will thus always be empty.  You need to edit your configuration before the database is used.  The reason is that storing file identifiers in plaintext in the database can compromise your privacy if your machine should fall under the control of an adversary.

.TP
\fB\-a, \fB\-\-list-all\fR
display all entries from the file-identifier database
.TP
\fB\-A, \fB\-\-kill-all\fR
delete all entries from the file-identifier database
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-i, \fB\-\-list-insert\fR
display entries from the insertion category in the file-identifier database
.TP
\fB\-I, \fB\-\-kill-insert\fR
delete all entries from the insertion category in the file-identifier database
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=LOGLEVEL\fR
change the loglevel. Possible values for LOGLEVEL are NOTHING, FATAL, ERROR, FAILURE, WARNING, MESSAGE, INFO, DEBUG, CRON and EVERYTHING.
.TP
\fB\-s, \fB\-\-list-search\fR
display entries from the search-result category in the file-identifier database
.TP
\fB\-S, \fB\-\-kill-search\fR
delete all entries from the search-result category in the file-identifier database
.TP
\fB\-x, \fB\-\-list-directory\fR
display entries from the directory-contents category in the file-identifier databae
.TP
\fB\-X, \fB\-\-kill-directory\fR
delete all entries from the directory-contents category in the file-identifier databae
.SH NOTES

In order to allow users maximum flexibility when building directories, GNUnet collects file-identifiers of files that the clients encounter in the file-identifier database.  The file-identifiers are classified into four categories (search, insert, directory, namespace) according to how they were obtained (as search results, by insertion of that file, as part of a directory that was downloaded).  When a user uses gnunet-gtk to build a directory, he can build the directory using a subset of the file-identifiers stored in the file-identifier database.

A GNUnet directory is a file containing a list of keys.  The keys can point to files, other directories or files in namespaces.  In other words, a GNUnet directory is similar to UNIX directories.  The difference to tar and zip is that GNUnet directory does not contain the actual files, just symbolic (links), similar to directories with symbolic links in UNIX filesystems.  The benefit is that the individual files can be retrieved separately (if desired) and if some of the files are inserted to another node in GNUnet, this just increases their availability but does not produce useless duplicates (for example, it is a better idea to publish a collection of pictures or compressed sound files using a GNUnet directory instead of processing them with archivers such as tar or zip first).

Directories can be created at the moment by \fBgnunet\-gtk\fP and \fBgnunet\-insert\fP.  They can point to content created by the user or content inserted by others.  Just like ordinary files, a directory can be published under a pseudonym-signed namespace.  A directory can also contain links to files in namespaces.

GNUnet-directories use the (unregistered) mimetype \fBapplication/gnunet\-directory\fP.  They can show up among normal search results or be pointed-to by sblocks. The directory file can be downloaded to disk by \fBgnunet\-download\fP(1) for later processing or be handled more directly by \fBgnunet\-gtk\fP(1).

.SH "REPORTING BUGS"
Report bugs by using mantis <http://www.ovmj.org/~mantis/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet\-gtk\fP(1), \fBgnunet\-insert\fP(1), \fBgnunet\-search\fP(1), \fBgnunet\-download\fP(1), \fBgnunetd\fP(1)


ChangeLog
Mon Feb 28 00:33:03 EST 2005
	Releasing GNUnet 0.6.6b.

Sun Feb 27 19:27:44 EST 2005
	Fixing mantis 839.

Sun Feb 27 18:44:11 EST 2005
	Fixing mantis 833/834.

Sun Feb 27 18:15:02 EST 2005
	Fixing serious problem with LE in gnunet-insert.
	Fixing mantis 828.

Thu Feb 17 01:14:30 EST 2005
	Various memory leaks for Win32 fixed.
	Added support for GNUNET(D)_CONFIG environment variable.

Wed Feb  9 13:33:38 EST 2005
        Fixed problems with the logrotate code.  Releasing GNUnet 0.6.6a.

Sun Jan 11 19:58:13 CET 2005
	Releasing GNUnet 0.6.6

Sun Jan 11 19:24:27 CET 2005
        gnunet-gtk now resumes aborted downloads automatically.

Sun Nov 14 16:49:02 EST 2004
	Debugged collections, seems to work (not that it looks pretty,
	but it basically works, the rest is UI work).  Fixed deadlock
	in gnunet-gtk logging.  Releasing GNUnet 0.6.5.

Thu Nov  4 09:04:48 EST 2004
	Added (highly experimental) support for collections.

Sat Sep 25 22:29:48 EST 2004
	Releasing GNUnet 0.6.4a.

Sun Sep 19 23:39:04 EST 2004
	Various code cleanups and minor bugfixes.

Sun Aug 29 15:11:28 IST 2004
	Releasing GNUnet 0.6.4.

Tue Aug 24 20:59:14 IST 2004
	Significant enhancements to namespaces.  A namespace can
	now be annotated with meta-data at the time that it is
	created.
	
Sat Aug 21 01:00:06 IST 2004
	Adding GNU gettext support to configure.
	
Thu Aug 19 01:26:55 IST 2004
	Migrating gnunet-gtk to GTK 2.0 (tested with 2.4).

Mon Aug 16 00:30:59 IST 2004
	Removed gcry.  Added dependency on GNU MP library (libgmp)
	for kblocks.

Tue Aug  3 17:57:43 IST 2004
	Fixed bugs in gnunet-update.  Fixed segfault in AFS startup.
	Fixed warnings in gnunet-peer-info/gnunetd.  Fixed packaging
	error (not all of util/gcry included in distribution).
	Releasing GNUnet 0.6.3a.

Mon Aug  2 20:06:07 IST 2004
	Releasing GNUnet 0.6.3.

Sun Aug  1 01:31:19 IST 2004
	Changed most places from HexName to the new, shorter EncName.
	Pushed URIs for namespaces and keyword searches into more
	widespread use.  Replaced all occurences of sprintf with
	SNPRINTF (to obtain extra bounds check).
	
Sun Jul 11 13:54:01 IST 2004
	Moved testcases for libgnunet_util to util/ directory.
	Made DB testcases in afs/module/ run for each available
	database and not just the one configured in /etc/gnunet.conf.

Fri Jun 25 21:05:59 IST 2004
	Releasing GNUnet 0.6.2b (resolves various minor bugs).

Wed May  5 17:34:39 EST 2004
	Releasing GNUnet 0.6.2a (resolves libgcrypt and packageing
	issues, improved gtk-statistics, other minor bugfixes).
	
Fri Apr 30 21:36:23 EST 2004
	Fixed synchronization problem in cron (only relevant for
	parallel downloading).  Releasing GNUnet 0.6.2.

Mon Apr 26 21:13:45 EST 2004
	Added support to gnunet-check to convert file index database
	to new design.  Removed support for pre 0.6.1b bloomfilter
	conversion (would have been too messy to support both).
	Added old-version check to gnunetd startup to make updates
	more systematic in the future (not perfect, though).

Mon Apr 26 19:27:29 EST 2004
	Recursive insertion of directories with gnunet-gtk 
	now works including the progress bar.

Sun Apr 25 18:24:55 EST 2004
	Global changes to fix bug #698.  The fix is still 
	incomplete with respect to pthread.h specific structs.
	
Sun Apr 25 15:20:29 EST 2004
	Fixed compile error in IPv6 transports.

Fri Apr 23 23:38:01 EST 2004
	Added recursive insertion of directories to gnunet-gtk.
	Progress-bar does not work yet, also libextractor is
	always used (no option to turn it off).
	
Fri Apr 23 21:49:18 EST 2004
	Removed need to specify _protocol in configuration file
	(shorter is better).

Wed Apr 21 23:32:36 EST 2004
	Added code to bound priority used by clients by twice the
	average priority of requests from other peers that are 
	routed at the moment.  That prevents clients from exposing
	themselves with excessively high priorities (and also 
	avoids giving too much credit/trust needlessly).

Tue Apr 20 22:27:19 EST 2004
	Added possibility of using a link instead of making a copy
	if gnunetd and the insert-client both run on the same machine.

Tue Apr 20 00:00:58 EST 2004
	Added network transparency for indexing files (by copying
	them to a GNUnet(d) specific directory).  This breaks
	unindexing (for files indexed with previous versions) and
	forces the user to make a copy of the file.  In the near
	future we should resolve this problem by allowing the use
	of a link.

Sun Apr 18 17:24:34 EST 2004
	Adding recursive, parallel download of directories to
	gnunet-download

Sun Apr 18 01:57:32 EST 2004
	gnunet-transport-check can now be used to check
	external connectivity (with other peers).
	
Sat Apr 17 17:46:19 EST 2004
	Reduced memory consuption for AFS insertion and
	download.
	
Sat Apr 10 21:17:46 EST 2004
	Implemented statistics-plot in gnunet-gtk based on
	gnome-system-monitor code.

Fri Apr  9 22:01:51 EST 2004
	Added support for libextractor > 0.2.7.
	
Fri Apr  9  0:29:20 EST 2004
	Completed --help conversion.

Wed Mar 31 15:40:51 EST 2004
	Releasing 0.6.1d.

Tue Mar 30 22:42:13 EST 2004
	Integrated gnunet-pseudonym-create, -delete and -list into
	gnunet-pseudonym as well as gnunet-directory-list, -emptydb
	and -print into gnunet-directory.

Tue Mar  2 10:46:01 EST 2004
	Fixing bug preventing download for indexed files <= 1k.

Sat Feb 28 03:43:34 EST 2004
	Releasing 0.6.1c.

Sat Feb 21 06:16:47 EST 2004
	Modified requestmanager to improve performance at the end of the 
	download.

Thu Feb 19 00:48:33 EST 2004
	Fixed integer underflow possibly causing slow download performance.
	
Sun Feb 15 15:17:05 EST 2004
	Fixed weak key failures with libgcrypt.

Sun Feb 15 13:14:44 EST 2004
	New routing code (untested) commited.

Sun Feb 15 11:58:10 EST 2004
	Fixing problem with 64-bit size_t on alpha (#655).

Sat Jan 31 04:32:33 EST 2004
	Releasing 0.6.1b.

Fri Jan 23 00:00:09 EST 2004
	Fixed indexed content being overwritten by migrated content
	
Thu Jan 22 19:27:33 EST 2004
	Fixed integer-overflow problems in statuscalls.

src/applications/afs/tools/gnunet-directory.c
/*
     This file is part of GNUnet.
     (C) 2002, 2003 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/afs/tools/gnunet-directory.c
 * @brief tool to list the entries stored in the database holding files for building directories,
 *        to delete all of these entries and to display the contents of
 *        directories.
 * @author Christian Grothoff
 **/

#include "gnunet_afs_esed2.h"
#include "platform.h"
  
static unsigned int listMask = 0;
static unsigned int killMask = 0;
static char ** filenames;
static int filenamescnt;


static void printNode(RootNode * root,
		      void * unused) {
  char * string;

  string = rootNodeToString(root);
  printf("%s\n", string);
  FREE(string);
}

static void printDirectory(char * filename) {
  GNUnetDirectory * dir;
  unsigned int i;

  filename = expandFileName(filename);
  printf("==> Directory '%s':\n",
	 filename);
  dir = readGNUnetDirectory(filename);
  if (dir == NULL) { 
    printf("ERROR: no such file or invalid format for GNUnet directory.\n");
    return;
  }
  dir->description[MAX_DESC_LEN-1] = '\0';
  printf("%s\n",
	 dir->description);
  for (i=0;i<ntohl(dir->number_of_files);i++) {
    printf("%u: ", i);
    printNode(&((GNUnetDirectory_GENERIC*)dir)->contents[i], NULL);
  }     
  printf("\n");
  FREE(dir);
  FREE(filename);
}

/**
 * Print a list of the options we offer.
 **/
static void printhelp() {
  static Help help[] = {
    { 'a', "list-all", NULL,
      "list all entries from the directory database" },
    { 'A', "kill-all", NULL,
      "remove all entries from the directory database" },    
    HELP_CONFIG,
    HELP_HELP,
    { 'i', "list-insert", NULL,
      "list all insert entries from the directory database" },
    { 'I', "kill-insert", NULL,
      "delete all insert entries from the directory database" },
    HELP_LOGLEVEL,
    { 'n', "list-namespace", NULL,
      "list all namespace entries from the directory database" },
    { 'N', "kill-namespace", NULL,
      "delete all namespace entries from the directory database" },
    { 's', "list-search", NULL,
      "list all search result entries from the directory database" },
    { 'S', "kill-search", NULL,
      "delete all search result entries from the directory database" },
    HELP_VERSION,
    { 'x', "list-directory", NULL,
      "list all directory entries from the directory database" },
    { 'X', "kill-directory", NULL,
      "remove all directory entries from the directory database" },
    HELP_END,
  };
  formatHelp("gnunet-directory [OPTIONS] [FILENAMES]",
	     "Perform directory related operations.",
	     help);
}

/**
 * Perform option parsing from the command line. 
 **/
static int parseCommandLine(int argc, 
			    char * argv[]) {
  int c;

  while (1) {
    int option_index = 0;
    static struct GNoption long_options[] = {
      LONG_DEFAULT_OPTIONS,
      { "list-search",    0, 0, 's' },
      { "list-insert",    0, 0, 'i' },
      { "list-directory",    0, 0, 'x' },
      { "list-namespace",    0, 0, 'n' },
      { "kill-search",    0, 0, 'S' },
      { "kill-insert",    0, 0, 'I' },
      { "kill-directory",    0, 0, 'X' },
      { "kill-namespace",    0, 0, 'N' },
      { "list-all",    0, 0, 'a' },
      { "kill-all",    0, 0, 'A' },
      { 0,0,0,0 }
    };
    
    c = GNgetopt_long(argc,
		      argv, 
		      "vhdc:L:sixanSIXAN", 
		      long_options, 
		      &option_index);    
    if (c == -1) 
      break;  /* No more flags to process */
    if (YES == parseDefaultOptions(c, GNoptarg))
      continue;    
    switch(c) {
    case 'a':
      listMask = DIR_CONTEXT_ALL;
      break;
    case 's':
      listMask = DIR_CONTEXT_SEARCH;
      break;
    case 'i':
      listMask = DIR_CONTEXT_INSERT;
      break;
    case 'n':
      listMask = DIR_CONTEXT_INSERT_SB;
      break;
    case 'x':
      listMask = DIR_CONTEXT_DIRECTORY;
      break;
    case 'A':
      killMask = DIR_CONTEXT_ALL;
      break;
    case 'S':
      killMask = DIR_CONTEXT_SEARCH;
      break;
    case 'I':
      killMask = DIR_CONTEXT_INSERT;
      break;
    case 'N':
      killMask = DIR_CONTEXT_INSERT_SB;
      break;
    case 'X':
      killMask = DIR_CONTEXT_DIRECTORY;
      break;
    case 'v': 
      printf("GNUnet v%s, gnunet-directory v%s\n",
	     VERSION, 
	     AFS_VERSION);
      return SYSERR;
    case 'h': 
      printhelp(); 
      return SYSERR;
    default:
      printf("Unknown option %c. Aborting.\n"
	     "Use --help to get a list of options.\n",
	     c);
      return SYSERR;
    } /* end of parsing commandline */
  }
  filenames = &argv[GNoptind];
  filenamescnt = argc - GNoptind;
  return OK;
}

int main(int argc, 
	 char * argv[]) {
  int i;
  if (SYSERR == initUtil(argc, argv, &parseCommandLine))
    return 0;

  if (listMask != 0)
    printf("Listed %d matching entries.\n",
	   iterateDirectoryDatabase(listMask,
				    &printNode,
				    NULL));
  if (killMask != 0)
    emptyDirectoryDatabase(killMask);  
  for (i=0;i<filenamescnt;i++)
    printDirectory(filenames[i]);
 
  doneUtil();
  return 0;
}

/* end of gnunet-directory.c */
